<font size="5">**Project: Demographics**

## Introduction

For most of human history, fertility rates have remained high—well above the 2.1 children per woman needed for replacement. Families were large. This is because, in times where demanding agricultural work was the norm, so too was childhood mortality. The more children a family had, the more likely it was that one or more would survive. However, as education and technology have increased, an interesting demographic phenomenon has emerged. The growth of education and technology has coincided with a decline in family sizes and fertility rates well below the replacement rate. Japan has a fertility rate of 1.39 children per woman, the U.S. 1.84, and Spain 1.29. By contrast, countries like Niger, Angola, and the DRC have fertility rates of 6.73, 5.76, and 5.56, respectively. That is an astonishing difference. It is something not limited to just European or Western countries but also countries like Japan and Taiwan. What is causing this, and why? What could this mean for the future? Such a colossal demographic change could have far-reaching consequences.

In 1979, the People’s Republic of China introduced the infamous One-Child Policy. Now, China has started experiencing the consequences of the so-called demographic transition. Most noticeably, the population pyramid has inverted. A once growing and young population is now dying and aging. Already, this has become noticeable in the United States. A glance at the population growth rate shows a plateauing curve, while the average age continues to climb. Still, why should we care? Health insurance premiums and health related expenses could quickly outpace incomes. Social Security, the delicately financed house of cards that it is, could collapse and leave tens of millions of senior citizens and disabled Americans without any support. Moreover, fewer workers might mean higher wages, but also less specialized labor and less access to vital services. Perhaps most importantly, shrinking and altogether disappearing families could have the direst consequences as support systems fall apart and an already atomized society loses its last institution.



## Research questions

...

## Data

The raw data for this project come from the World Bank, at [https://databank.worldbank.org/source/world-development-indicators/](https://databank.worldbank.org/source/world-development-indicators/).

It contains the following variables ...

**Note:** Create a CSV file with the data description (see my `datadescr.csv` example) and then use [https://www.tablesgenerator.com/markdown_tables](https://www.tablesgenerator.com/markdown_tables) to import it and convert to Markdown.

| Variable | Definition              | Units/Key              |
|----------|-------------------------|------------------------|
| Country  | Name of country         |                        |
| Debt     | Central government debt | % of GDP               |
| CO2      | CO2 emissions           | kg per 2010 US\$ of GDP |

### Data discussion

Discuss anything special about the data that the reader will need to know to understand your later analysis.

### Data cleaning

- Import pandas

In [1]:
import pandas as pd

- Read in raw demographics data

In [2]:
df = pd.read_csv('data/Demographics.csv')
df.head()

Country Name Country Code    Time Time Code  \
0  Afghanistan          AFG  2013.0    YR2013   
1  Afghanistan          AFG  2014.0    YR2014   
2  Afghanistan          AFG  2015.0    YR2015   
3  Afghanistan          AFG  2016.0    YR2016   
4  Afghanistan          AFG  2017.0    YR2017   

   Age dependency ratio, young (% of working-age population) [SP.POP.DPND.YG]  \
0                                          92.388046                            
1                                          90.015900                            
2                                          88.398202                            
3                                          87.405774                            
4                                          85.970407                            

  Fertility rate, total (births per woman) [SP.DYN.TFRT.IN]  \
0                                              5.696          
1                                               5.56          
2                                              5.405          
3                                              5.262          
4                                              5.129          

  GDP growth (annual %) [NY.GDP.MKTP.KD.ZG]  \
0                          5.60074465808154   
1                          2.72454336394854   
2                          1.45131466009755   
3                          2.26031420130452   
4                          2.64700320195786   

  Life expectancy at birth, total (years) [SP.DYN.LE00.IN]  \
0                                             62.417         
1                                             62.545         
2                                             62.659         
3                                             63.136         
4                                             63.016         

  School enrollment, primary (% gross) [SE.PRM.ENRR]  \
0                                   107.695976257324   
1                                   109.115516662598   
2                                   106.182418823242   
3                                   106.150283813477   
4                                   106.129997253418   

  Literacy rate, adult total (% of people ages 15 and above) [SE.ADT.LITR.ZS]  \
0                                                 ..                            
1                                                 ..                            
2                                                 ..                            
3                                                 ..                            
4                                                 ..                            

  Rural population (% of total population) [SP.RUR.TOTL.ZS]  
0                                             75.627         
1                                             75.413         
2                                             75.197         
3                                              74.98         
4                                              74.75

- Rename variables for ease of use

In [3]:
df = df.rename(columns={"Country Name":'Country', 
                        "Time":'Year', 
                        "Age dependency ratio, young (% of working-age population) [SP.POP.DPND.YG]":'Age Dependancy Ratio',
                        "Fertility rate, total (births per woman) [SP.DYN.TFRT.IN]": 'Fertility Rate',
                        "GDP growth (annual %) [NY.GDP.MKTP.KD.ZG]": 'GDP Growth',
                        "Life expectancy at birth, total (years) [SP.DYN.LE00.IN]": 'Life Expectancy',
                        "School enrollment, primary (% gross) [SE.PRM.ENRR]": 'Primary School Enrollment', 
                        "Literacy rate, adult total (% of people ages 15 and above) [SE.ADT.LITR.ZS]": 'Literacy Rate',
                        "Rural population (% of total population) [SP.RUR.TOTL.ZS]": 'Rural Population'})
df.head()

Country Country Code    Year Time Code  Age Dependancy Ratio  \
0  Afghanistan          AFG  2013.0    YR2013             92.388046   
1  Afghanistan          AFG  2014.0    YR2014             90.015900   
2  Afghanistan          AFG  2015.0    YR2015             88.398202   
3  Afghanistan          AFG  2016.0    YR2016             87.405774   
4  Afghanistan          AFG  2017.0    YR2017             85.970407   

  Fertility Rate        GDP Growth Life Expectancy Primary School Enrollment  \
0          5.696  5.60074465808154          62.417          107.695976257324   
1           5.56  2.72454336394854          62.545          109.115516662598   
2          5.405  1.45131466009755          62.659          106.182418823242   
3          5.262  2.26031420130452          63.136          106.150283813477   
4          5.129  2.64700320195786          63.016          106.129997253418   

  Literacy Rate Rural Population  
0            ..           75.627  
1            ..           75.413  
2            ..           75.197  
3            ..            74.98  
4            ..            74.75

- Filter out country code and time code

In [4]:
df = df.filter(['Country','Age Dependancy Ratio','Fertility Rate', 'GDP Growth', 
                'Life Expectancy', 'Primary School Enrollment', 'Literacy Rate', 'Rural Population'])
df.head()

Country  Age Dependancy Ratio Fertility Rate        GDP Growth  \
0  Afghanistan             92.388046          5.696  5.60074465808154   
1  Afghanistan             90.015900           5.56  2.72454336394854   
2  Afghanistan             88.398202          5.405  1.45131466009755   
3  Afghanistan             87.405774          5.262  2.26031420130452   
4  Afghanistan             85.970407          5.129  2.64700320195786   

  Life Expectancy Primary School Enrollment Literacy Rate Rural Population  
0          62.417          107.695976257324            ..           75.627  
1          62.545          109.115516662598            ..           75.413  
2          62.659          106.182418823242            ..           75.197  
3          63.136          106.150283813477            ..            74.98  
4          63.016          106.129997253418            ..            74.75

- Make a copy of current data frame

In [5]:
df2 = df.copy()

- Convert the Country variable in the dataframe to a series called Country

In [6]:
Country = df['Country']
Country.head()

0    Afghanistan
1    Afghanistan
2    Afghanistan
3    Afghanistan
4    Afghanistan
Name: Country, dtype: object

In [7]:
type(Country)

pandas.core.series.Series

- Change all non-numeric objects to NaN

In [8]:
for col in  df2.columns[1:]:
    df2[col] = pd.to_numeric(df2[col], errors='coerce')

In [9]:
df2.head()

Country  Age Dependancy Ratio  Fertility Rate  GDP Growth  \
0  Afghanistan             92.388046           5.696    5.600745   
1  Afghanistan             90.015900           5.560    2.724543   
2  Afghanistan             88.398202           5.405    1.451315   
3  Afghanistan             87.405774           5.262    2.260314   
4  Afghanistan             85.970407           5.129    2.647003   

   Life Expectancy  Primary School Enrollment  Literacy Rate  Rural Population  
0           62.417                 107.695976            NaN            75.627  
1           62.545                 109.115517            NaN            75.413  
2           62.659                 106.182419            NaN            75.197  
3           63.136                 106.150284            NaN            74.980  
4           63.016                 106.129997            NaN            74.750

Count non-missing observations for each variable

In [10]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1958 entries, 0 to 1957
Data columns (total 8 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Country                    1955 non-null   object 
 1   Age Dependancy Ratio       1953 non-null   float64
 2   Fertility Rate             1892 non-null   float64
 3   GDP Growth                 1867 non-null   float64
 4   Life Expectancy            1886 non-null   float64
 5   Primary School Enrollment  1494 non-null   float64
 6   Literacy Rate              309 non-null    float64
 7   Rural Population           1935 non-null   float64
dtypes: float64(7), object(1)
memory usage: 122.5+ KB


- Calculate the mean of all non-missing values by country

In [11]:
df = df2.groupby('Country').mean()
df.head()

Age Dependancy Ratio  Fertility Rate  GDP Growth  \
Country                                                            
Afghanistan                85.939664        5.146333   -0.362928   
Albania                    25.747409        1.518889    2.647364   
Algeria                    45.839689        2.993889    1.700000   
American Samoa             46.999079             NaN   -0.096902   
Andorra                    20.279530             NaN    0.572704   

                Life Expectancy  Primary School Enrollment  Literacy Rate  \
Country                                                                     
Afghanistan           62.775111                 107.580319      37.266041   
Albania               78.333222                 107.653550            NaN   
Algeria               75.576889                 109.901002      81.407837   
American Samoa              NaN                        NaN            NaN   
Andorra                     NaN                  89.001746            NaN   

                Rural Population  
Country                           
Afghanistan            74.708667  
Albania                40.702556  
Algeria                28.003556  
American Samoa         12.792889  
Andorra                11.828778

- Drop na's 

In [12]:
df = df.dropna()
df.head()

Age Dependancy Ratio  Fertility Rate  GDP Growth  \
Country                                                         
Afghanistan             85.939664        5.146333   -0.362928   
Algeria                 45.839689        2.993889    1.700000   
Angola                  87.853405        5.612556    0.170649   
Armenia                 29.046853        1.598889    3.233333   
Aruba                   26.537023        1.734111    2.447594   

             Life Expectancy  Primary School Enrollment  Literacy Rate  \
Country                                                                  
Afghanistan        62.775111                 107.580319      37.266041   
Algeria            75.576889                 109.901002      81.407837   
Angola             61.252222                 109.244814      66.030113   
Armenia            74.051000                  99.460075      99.756363   
Aruba              75.678778                 117.817928      97.989998   

             Rural Population  
Country                        
Afghanistan         74.708667  
Algeria             28.003556  
Angola              35.206333  
Armenia             36.815778  
Aruba               56.650667

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 123 entries, Afghanistan to Zimbabwe
Data columns (total 7 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Age Dependancy Ratio       123 non-null    float64
 1   Fertility Rate             123 non-null    float64
 2   GDP Growth                 123 non-null    float64
 3   Life Expectancy            123 non-null    float64
 4   Primary School Enrollment  123 non-null    float64
 5   Literacy Rate              123 non-null    float64
 6   Rural Population           123 non-null    float64
dtypes: float64(7)
memory usage: 7.7+ KB


- We have 123 countries left to work with, and here is our clean data

In [14]:
df

Age Dependancy Ratio  Fertility Rate  GDP Growth  \
Country                                                                
Afghanistan                    85.939664        5.146333   -0.362928   
Algeria                        45.839689        2.993889    1.700000   
Angola                         87.853405        5.612556    0.170649   
Armenia                        29.046853        1.598889    3.233333   
Aruba                          26.537023        1.734111    2.447594   
...                                  ...             ...         ...   
Venezuela, RB                  44.201793        2.296556   -1.275646   
Viet Nam                       33.369352        1.944889    5.871594   
West Bank and Gaza             71.026831        3.814667    1.869919   
Zambia                         83.675509        4.634778    3.027344   
Zimbabwe                       77.194439        3.738444    1.223848   

                    Life Expectancy  Primary School Enrollment  Literacy Rate  \
Country                                                                         
Afghanistan               62.775111                 107.580319      37.266041   
Algeria                   75.576889                 109.901002      81.407837   
Angola                    61.252222                 109.244814      66.030113   
Armenia                   74.051000                  99.460075      99.756363   
Aruba                     75.678778                 117.817928      97.989998   
...                             ...                        ...            ...   
Venezuela, RB             72.093889                  98.563437      96.866154   
Viet Nam                  74.052444                 111.475478      95.753868   
West Bank and Gaza        74.261556                  95.364152      96.710934   
Zambia                    61.604111                 100.721525      87.500000   
Zimbabwe                  59.999222                  99.378456      88.693420   

                    Rural Population  
Country                               
Afghanistan                74.708667  
Algeria                    28.003556  
Angola                     35.206333  
Armenia                    36.815778  
Aruba                      56.650667  
...                              ...  
Venezuela, RB              11.796556  
Viet Nam                   64.776111  
West Bank and Gaza         24.089778  
Zambia                     57.002000  
Zimbabwe                   67.662222  

[123 rows x 7 columns]

## Data analysis

Create histograms of key variables, and scatterplots to illustrate relationships between them, by adapting my examples in `exer04_ans.ipynb`.